### ISSUE TITLE
Improve error message and path

### ISSUE THREAD

**Issue by DavidBadura** [CuyZ/Valinor#329]
Thanks for this library, mapping data to strict objects is amazing!

I'm using Valinor for a rest API. However, it is currently difficult for the API user to return a proper error message.

I have the following objects:

```php
readonly class Foo {
  public function __construct(
    public Bar $bar
  ) {}
}

readonly class Bar {
  public function __construct(
    /** @var no-empty-string */
    public string $name
  ) {}
}
```

This data:

```json
{
  "bar": {
    "name": null
  }
}
```

And so I try to map the whole thing for a nice response:

```php
$messages = Messages::flattenFromNode(
    $mappingError->node()
);

$errors = [];

foreach ($messages as $message) {
    $errors[] = [
        'path' => $message->node()->path(),
        'message' => $message->toString(),
    ];
}

return $errors;
```

Current Result:

```json
[
  {
    "path": "*root*",
    "message": "Value array{bar: array{…}} is not a valid non-empty string."
  }
]
```

This isn't really helpful for the user and I don't know how to change it or if it's even possible currently. I would like to build the following output:

```json
[
  {
    "path": "bar.name",
    "message": "Value \"null\" is not a valid non-empty string."
  }
]
```
Is it already possible to generate this output? If not, is it something that could be changed? I can also try to create a PR.

**Comment by DanielBadura** [comment]
This is mainly caused by the changes made in https://github.com/CuyZ/Valinor/pull/256. So with version `1.0.0` of this library our integration tests fail. If I change `ClassNodeBuilder` on line `29` like this:
```diff
-        $node = count($children) === 1
-            ? TreeNode::flattenedBranch($shell, $object, $children[0])
-            : TreeNode::branch($shell, $object, $children);
+        $node = TreeNode::branch($shell, $object, $children);
```

Then the `path` is correct in our intergration tests. But then some integration tests of this library are failing. @romm maybe you have an idea how we could fix this? I think this is a change which was not intended, right?

**Comment by romm** [comment]
Hi there, you are indeed right @DanielBadura (I'd add #318 to the list of changes bound to this issue).

I definitely have to do something about it, but I think it will be a bit more complex than the change you suggested. Should be pretty easy though, I'll take a look when I have some time for this.

Sorry for the inconvenience; I'll keep you updated here!

**Comment by romm** [comment]
Hey there, thanks again for the report and investigation.

The issue should be fixed with #345 and will be part of the next release. 😉
